File: server/src/modules/patient/ml-prediction.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { CreatePredictionDto, PredictionResultDto } from './dto/create-prediction.dto';
import { SupabaseService } from '../supabase/supabase.service';

@Injectable()
export class MLPredictionService {
  private readonly logger = new Logger(MLPredictionService.name);

  // Age ranges and risk multipliers from the Python model
  private readonly ageRanges = {
    'Young_Adult': [18, 35],
    'Middle_Adult': [36, 50], 
    'Mature_Adult': [51, 65],
    'Senior': [66, 80],
    'Elderly': [81, 95]
  };

  private readonly ageRiskMultipliers = {
    'Young_Adult': 1.0,
    'Middle_Adult': 1.2,
    'Mature_Adult': 1.4,
    'Senior': 1.7,
    'Elderly': 2.1
  };

  // Clinical risk weights from the Python model
  private readonly clinicalWeights = {
    age_elderly: 0.12,
    age_senior: 0.08,
    diabetes: 0.10,
    hypertension: 0.08,
    heart_disease: 0.12,
    kidney_disease: 0.15,
    respiratory_disease: 0.12,
    icu_admission: 0.20,
    semi_intensive_admission: 0.15,
    regular_ward_admission: 0.05,
    long_stay: 0.10,
    low_hemoglobin: 0.08,
    abnormal_hematocrit: 0.06,
    low_platelets: 0.10,
    abnormal_rbc: 0.06,
    low_lymphocytes: 0.08,
    high_urea: 0.10,
    electrolyte_imbalance: 0.08,
    covid_positive: 0.15,
    frequent_admissions: 0.12,
    high_medications: 0.08,
    multiple_comorbidities: 0.18
  };

  constructor(private readonly supabaseService: SupabaseService) {}

  /**
   * Standardize age group based on patient age
   */
  private standardizeAgeGroup(age: number): string {
    if (!age || age < 18) return 'Unknown';
    
    for (const [group, [min, max]] of Object.entries(this.ageRanges)) {
      if (age >= min && age <= max) {
        return group;
      }
    }
    return 'Unknown';
  }

  /**
   * Calculate comprehensive clinical risk score
   */
  private calculateClinicalScore(patient: any, predictionData: CreatePredictionDto): { score: number, riskFactors: string[] } {
    let riskScore = 0.0;
    const riskFactors: string[] = [];

    // Age factors
    const ageGroup = this.standardizeAgeGroup(patient.age || this.calculateAge(patient.dob));
    if (ageGroup === 'Elderly') {
      riskScore += this.clinicalWeights.age_elderly;
      riskFactors.push('Advanced age (80+ years) - very high risk');
    } else if (ageGroup === 'Senior') {
      riskScore += this.clinicalWeights.age_senior;
      riskFactors.push('Senior age (65-80 years) - increased risk');
    }

    // Chronic conditions
    const conditions = {
      diabetes: 'Diabetes mellitus',
      hypertension: 'Hypertension', 
      heart_disease: 'Heart disease',
      kidney_disease: 'Kidney disease',
      respiratory_disease: 'Respiratory disease'
    };

    const activeConditions: string[] = [];
    for (const [condition, label] of Object.entries(conditions)) {
      if (predictionData[condition] === 1) {
        riskScore += this.clinicalWeights[condition] || 0.08;
        activeConditions.push(label);
      }
    }

    if (activeConditions.length >= 3) {
      riskScore += this.clinicalWeights.multiple_comorbidities;
      riskFactors.push(`Multiple comorbidities: ${activeConditions.join(', ')}`);
    } else if (activeConditions.length > 0) {
      riskFactors.push(...activeConditions);
    }

    // Hospital admission severity
    if (predictionData.intensive_care_unit_admission === 1) {
      riskScore += this.clinicalWeights.icu_admission;
      riskFactors.push('ICU admission - critical condition');
    } else if (predictionData.semi_intensive_unit_admission === 1) {
      riskScore += this.clinicalWeights.semi_intensive_admission;
      riskFactors.push('Semi-intensive unit admission');
    }

    // Lab abnormalities
    if (predictionData.hemoglobin < 12) {
      riskScore += this.clinicalWeights.low_hemoglobin;
      riskFactors.push(`Low hemoglobin (anemia): ${predictionData.hemoglobin} g/dL`);
    }

    if (predictionData.hematocrit < 35 || predictionData.hematocrit > 50) {
      riskScore += this.clinicalWeights.abnormal_hematocrit;
      riskFactors.push(`Abnormal hematocrit: ${predictionData.hematocrit}%`);
    }

    if (predictionData.platelets < 150) {
      riskScore += this.clinicalWeights.low_platelets;
      riskFactors.push(`Low platelets (thrombocytopenia): ${predictionData.platelets} x10³/μL`);
    }

    if (predictionData.red_blood_cells < 4.0 || predictionData.red_blood_cells > 6.0) {
      riskScore += this.clinicalWeights.abnormal_rbc;
      riskFactors.push(`Abnormal RBC count: ${predictionData.red_blood_cells} x10⁶/μL`);
    }

    if (predictionData.lymphocytes < 1.0) {
      riskScore += this.clinicalWeights.low_lymphocytes;
      riskFactors.push(`Low lymphocytes (immunocompromised): ${predictionData.lymphocytes} x10³/μL`);
    }

    if (predictionData.urea > 7.5) {
      riskScore += this.clinicalWeights.high_urea;
      riskFactors.push(`Elevated urea (kidney dysfunction): ${predictionData.urea} mmol/L`);
    }

    // Electrolyte imbalance
    if (predictionData.potassium < 3.5 || predictionData.potassium > 5.0 || 
        predictionData.sodium < 136 || predictionData.sodium > 145) {
      riskScore += this.clinicalWeights.electrolyte_imbalance;
      riskFactors.push(`Electrolyte imbalance (K: ${predictionData.potassium}, Na: ${predictionData.sodium})`);
    }

    // COVID-19
    if (predictionData.sars_cov2_exam_result === 1) {
      riskScore += this.clinicalWeights.covid_positive;
      riskFactors.push('COVID-19 positive');
    }

    // Length of stay
    if (predictionData.length_of_stay > 10) {
      riskScore += this.clinicalWeights.long_stay;
      riskFactors.push(`Extended hospitalization (${predictionData.length_of_stay} days)`);
    }

    // Frequent admissions
    if (predictionData.previous_admissions >= 2) {
      riskScore += this.clinicalWeights.frequent_admissions;
      riskFactors.push(`Frequent admissions (${predictionData.previous_admissions} in past year)`);
    }

    // High medication count
    if (predictionData.num_medications >= 10) {
      riskScore += this.clinicalWeights.high_medications;
      riskFactors.push(`Polypharmacy (${predictionData.num_medications} medications)`);
    }

    return { score: Math.min(riskScore, 1.0), riskFactors };
  }

  /**
   * Simulate ML model prediction (simplified version of the complex Python models)
   */
  private simulateMLPrediction(patient: any, predictionData: CreatePredictionDto): number {
    // This is a simplified version of the ML model from the Python code
    // In a real implementation, you would call the actual trained models
    
    let mlScore = 0.25; // Base probability

    // Age contribution
    const age = patient.age || this.calculateAge(patient.dob);
    if (age > 65) mlScore += 0.15;
    if (age > 80) mlScore += 0.10;

    // Comorbidity contribution
    const comorbidityCount = predictionData.diabetes + predictionData.hypertension + 
                           predictionData.heart_disease + predictionData.kidney_disease + 
                           predictionData.respiratory_disease;
    mlScore += comorbidityCount * 0.08;

    // ICU admission is a strong predictor
    if (predictionData.intensive_care_unit_admission === 1) {
      mlScore += 0.25;
    }

    // Lab values contribution
    if (predictionData.hemoglobin < 10) mlScore += 0.12;
    if (predictionData.platelets < 100) mlScore += 0.15;
    if (predictionData.urea > 15) mlScore += 0.18;

    // Length of stay
    if (predictionData.length_of_stay > 14) mlScore += 0.10;

    // Previous admissions
    mlScore += predictionData.previous_admissions * 0.05;

    return Math.min(mlScore, 0.95);
  }

  /**
   * Check for critical medical conditions
   */
  private hasCriticalConditions(predictionData: CreatePredictionDto): boolean {
    return predictionData.intensive_care_unit_admission === 1 ||
           predictionData.hemoglobin < 8 ||
           predictionData.platelets < 50 ||
           predictionData.urea > 20 ||
           predictionData.length_of_stay > 20 ||
           (predictionData.diabetes + predictionData.hypertension + 
            predictionData.heart_disease + predictionData.kidney_disease + 
            predictionData.respiratory_disease) >= 4;
  }

  /**
   * Get risk level from probability
   */
  private getRiskLevel(probability: number): string {
    if (probability >= 0.55) return 'high';
    if (probability >= 0.25) return 'medium';
    return 'low';
  }

  /**
   * Get clinical recommendation
   */
  private getRecommendation(riskLevel: string, probability: number): string {
    if (riskLevel === 'high') {
      if (probability >= 0.75) {
        return "🚨 URGENT: Comprehensive discharge planning, intensive case management, transitional care team, 24h post-discharge contact, consider delaying discharge";
      } else {
        return "⚠️ HIGH RISK: Comprehensive discharge planning, case management, transitional care, 24-48h post-discharge contact";
      }
    } else if (riskLevel === 'medium') {
      return "📋 MODERATE RISK: Enhanced discharge planning, medication reconciliation, patient education, 3-7 day follow-up";
    } else {
      return "✅ LOW RISK: Standard discharge planning, routine follow-up in 1-2 weeks, patient education materials";
    }
  }

  /**
   * Calculate age from date of birth
   */
  private calculateAge(dob: string): number {
    if (!dob) return 50; // Default age if not provided
    const today = new Date();
    const birthDate = new Date(dob);
    let age = today.getFullYear() - birthDate.getFullYear();
    const monthDiff = today.getMonth() - birthDate.getMonth();
    if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birthDate.getDate())) {
      age--;
    }
    return age;
  }

  /**
   * Main prediction method
   */
  async predictReadmissionRisk(patient: any, predictionData: CreatePredictionDto): Promise<PredictionResultDto> {
    try {
      this.logger.log(`Starting ML prediction for patient: ${patient.id}`);

      // Calculate clinical risk score
      const { score: clinicalScore, riskFactors } = this.calculateClinicalScore(patient, predictionData);

      // Get ML prediction
      const mlProbability = this.simulateMLPrediction(patient, predictionData);

      // Get age group and multiplier
      const age = patient.age || this.calculateAge(patient.dob);
      const ageGroup = this.standardizeAgeGroup(age);
      const ageMultiplier = this.ageRiskMultipliers[ageGroup] || 1.0;

      // Combined prediction (60% ML, 40% clinical, with age multiplier)
      const baseProbability = (0.60 * mlProbability) + (0.40 * clinicalScore);
      let finalProbability = Math.min(baseProbability * ageMultiplier, 1.0);

      // Critical condition overrides
      if (this.hasCriticalConditions(predictionData)) {
        finalProbability = Math.max(finalProbability, 0.70);
        riskFactors.push('Critical medical conditions detected');
      }

      // ICU admission override
      if (predictionData.intensive_care_unit_admission === 1) {
        finalProbability = Math.max(finalProbability, 0.65);
        riskFactors.push('ICU admission - high risk');
      }

      const riskLevel = this.getRiskLevel(finalProbability);
      const recommendation = this.getRecommendation(riskLevel, finalProbability);

      // Calculate confidence (simplified)
      const confidence = Math.min(75 + (clinicalScore * 15), 95);

      // Store prediction in database
      const predictionResult = await this.storePrediction({
        patient_id: patient.id,
        doctor_id: predictionData.doctor_id,
        risk_level: riskLevel,
        risk_percentage: finalProbability * 100,
        ml_probability: mlProbability * 100,
        clinical_score: clinicalScore * 100,
        age_multiplier: ageMultiplier,
        risk_factors: riskFactors,
        recommendation,
        confidence,
        age_group: ageGroup,
        predicted_at: new Date().toISOString(),
        input_data: predictionData
      });

      this.logger.log(`ML prediction completed for patient: ${patient.id}, risk level: ${riskLevel}`);

      return predictionResult;

    } catch (error) {
      this.logger.error(`Error in ML prediction: ${error.message}`);
      throw error;
    }
  }

  /**
   * Store prediction result in database
   */
  private async storePrediction(predictionData: any): Promise<PredictionResultDto> {
    try {
      const supabase = this.supabaseService.getClient();

      const { data: savedPrediction, error } = await supabase
        .from('ml_predictions')
        .insert(predictionData)
        .select()
        .single();

      if (error) {
        this.logger.error(`Error storing prediction: ${error.message}`);
        throw new Error(`Failed to store prediction: ${error.message}`);
      }

      return savedPrediction;
    } catch (error) {
      this.logger.error(`Error storing prediction: ${error.message}`);
      throw error;
    }
  }

  /**
   * Get patient predictions history
   */
  async getPatientPredictions(patientId: string): Promise<PredictionResultDto[]> {
    try {
      const supabase = this.supabaseService.getClient();

      const { data: predictions, error } = await supabase
        .from('ml_predictions')
        .select('*')
        .eq('patient_id', patientId)
        .order('predicted_at', { ascending: false });

      if (error) {
        this.logger.error(`Error fetching predictions: ${error.message}`);
        throw new Error(`Failed to fetch predictions: ${error.message}`);
      }

      return predictions || [];
    } catch (error) {
      this.logger.error(`Error fetching predictions: ${error.message}`);
      throw error;
    }
  }

  /**
   * Get all predictions for a department
   */
  async getDepartmentPredictions(department?: string): Promise<any[]> {
    try {
      const supabase = this.supabaseService.getClient();

      let query = supabase
        .from('ml_predictions')
        .select(`
          *,
          patients!inner(
            first_name,
            last_name,
            mrn,
            department
          )
        `)
        .order('predicted_at', { ascending: false });

      if (department) {
        query = query.eq('patients.department', department);
      }

      const { data: predictions, error } = await query;

      if (error) {
        this.logger.error(`Error fetching department predictions: ${error.message}`);
        throw new Error(`Failed to fetch predictions: ${error.message}`);
      }

      return predictions || [];
    } catch (error) {
      this.logger.error(`Error fetching department predictions: ${error.message}`);
      throw error;
    }
  }
}

File: server/src/services/ml_prediction_service.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Simplified ML Prediction Service for RelayLoop
This service provides readmission predictions using the enhanced ML model
"""

import sys
import os
import json
import numpy as np
import pandas as pd
from typing import Dict, Any, Optional

# Try to import ML libraries, fall back to basic prediction if not available
try:
    from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, ExtraTreesClassifier
    from sklearn.linear_model import LogisticRegression
    from sklearn.preprocessing import StandardScaler, LabelEncoder, RobustScaler
    from sklearn.model_selection import train_test_split
    from sklearn.metrics import accuracy_score, roc_auc_score, f1_score
    import warnings
    warnings.filterwarnings('ignore')
    ML_AVAILABLE = True
except ImportError:
    ML_AVAILABLE = False

class SimplifiedMLService:
    """Simplified ML service for readmission prediction"""
    
    def __init__(self):
        self.is_initialized = False
        self.models = {}
        self.scalers = {}
        self.label_encoders = {}
        
        # Clinical risk weights
        self.clinical_weights = {
            'age_elderly': 0.12,
            'age_senior': 0.08,
            'diabetes': 0.10,
            'hypertension': 0.08,
            'heart_disease': 0.12,
            'kidney_disease': 0.15,
            'respiratory_disease': 0.12,
            'icu_admission': 0.20,
            'semi_intensive_admission': 0.15,
            'regular_ward_admission': 0.05,
            'long_stay': 0.10,
            'low_hemoglobin': 0.08,
            'low_platelets': 0.10,
            'high_urea': 0.10,
            'covid_positive': 0.15,
            'frequent_admissions': 0.12,
            'high_medications': 0.08,
            'multiple_comorbidities': 0.18
        }
        
        # Age multipliers
        self.age_multipliers = {
            'Young_Adult': 1.0,
            'Middle_Adult': 1.2,
            'Mature_Adult': 1.4,
            'Senior': 1.7,
            'Elderly': 2.1
        }
    
    def initialize(self, data_path: str = None):
        """Initialize the ML service"""
        try:
            if ML_AVAILABLE:
                # If we have CSV data, load and train
                if data_path and os.path.exists(data_path):
                    csv_files = [f for f in os.listdir(data_path) if f.endswith('.csv')]
                    if csv_files:
                        self._load_and_train_from_csv(data_path, csv_files[0])
                    else:
                        self._create_synthetic_training_data()
                else:
                    self._create_synthetic_training_data()
            
            self.is_initialized = True
            # Don't print initialization message to avoid JSON parsing issues
            
        except Exception as e:
            # Don't print errors to avoid JSON parsing issues
            self.is_initialized = True  # Still allow fallback predictions
    
    def _create_synthetic_training_data(self):
        """Create synthetic training data and train models"""
        if not ML_AVAILABLE:
            return
            
        # Create synthetic data
        np.random.seed(42)
        n_samples = 1000
        
        data = {
            'age': np.random.randint(18, 90, n_samples),
            'diabetes': np.random.choice([0, 1], n_samples, p=[0.7, 0.3]),
            'hypertension': np.random.choice([0, 1], n_samples, p=[0.6, 0.4]),
            'heart_disease': np.random.choice([0, 1], n_samples, p=[0.8, 0.2]),
            'kidney_disease': np.random.choice([0, 1], n_samples, p=[0.85, 0.15]),
            'respiratory_disease': np.random.choice([0, 1], n_samples, p=[0.8, 0.2]),
            'intensive_care_unit_admission': np.random.choice([0, 1], n_samples, p=[0.9, 0.1]),
            'hemoglobin': np.random.normal(13.5, 2.0, n_samples).clip(6, 18),
            'platelets': np.random.normal(250, 80, n_samples).clip(50, 600),
            'urea': np.random.normal(5.0, 3.0, n_samples).clip(1, 20),
            'length_of_stay': np.random.exponential(5, n_samples).clip(1, 30),
            'sars_cov2_exam_result': np.random.choice([0, 1], n_samples, p=[0.9, 0.1]),
            'previous_admissions': np.random.poisson(1, n_samples),
            'num_medications': np.random.poisson(6, n_samples)
        }
        
        df = pd.DataFrame(data)
        
        # Create target variable based on risk factors
        risk_score = (
            (df['age'] > 65) * 0.2 +
            df['diabetes'] * 0.1 +
            df['hypertension'] * 0.08 +
            df['heart_disease'] * 0.12 +
            df['kidney_disease'] * 0.15 +
            df['intensive_care_unit_admission'] * 0.2 +
            (df['hemoglobin'] < 12) * 0.1 +
            (df['length_of_stay'] > 10) * 0.1 +
            df['sars_cov2_exam_result'] * 0.15
        )
        
        # Convert to binary outcome with some noise
        readmission_prob = 1 / (1 + np.exp(-3 * (risk_score - 0.5)))
        df['readmitted'] = np.random.binomial(1, readmission_prob)
        
        self._train_models(df)
    
    def _load_and_train_from_csv(self, data_path: str, csv_file: str):
        """Load data from CSV and train models"""
        if not ML_AVAILABLE:
            return
            
        try:
            csv_path = os.path.join(data_path, csv_file)
            df = pd.read_csv(csv_path)
            
            # Basic preprocessing - map common column names
            column_mappings = {
                'readmitted_30_days': 'readmitted',
                'patient_age': 'age',
                'sars_cov2_exam_result': 'covid_result'
            }
            
            df = df.rename(columns=column_mappings)
            
            # Ensure we have a target column
            if 'readmitted' not in df.columns:
                if 'readmitted_30_days' in df.columns:
                    df['readmitted'] = df['readmitted_30_days']
                else:
                    # Create synthetic target based on available features
                    df['readmitted'] = np.random.choice([0, 1], len(df), p=[0.8, 0.2])
            
            self._train_models(df)
            # Don't print success message to avoid JSON parsing issues
            
        except Exception as e:
            # Don't print errors to avoid JSON parsing issues
            pass
    
    def _train_models(self, df: pd.DataFrame):
        """Train ML models"""
        if not ML_AVAILABLE:
            return
            
        try:
            # Select features
            feature_cols = []
            for col in ['age', 'diabetes', 'hypertension', 'heart_disease', 'kidney_disease', 
                       'respiratory_disease', 'intensive_care_unit_admission', 'hemoglobin', 
                       'platelets', 'urea', 'length_of_stay', 'sars_cov2_exam_result', 
                       'previous_admissions', 'num_medications']:
                if col in df.columns:
                    feature_cols.append(col)
            
            if not feature_cols or 'readmitted' not in df.columns:
                print("Insufficient data for training")
                return
            
            X = df[feature_cols].fillna(0)
            y = df['readmitted']
            
            if len(X) < 50:  # Insufficient data
                print("Insufficient training data")
                return
            
            X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
            
            # Train models
            self.scalers['standard'] = StandardScaler()
            X_train_scaled = self.scalers['standard'].fit_transform(X_train)
            
            models = {
                'random_forest': RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42, class_weight='balanced'),
                'logistic_regression': LogisticRegression(random_state=42, class_weight='balanced', max_iter=1000)
            }
            
            for name, model in models.items():
                try:
                    if name == 'logistic_regression':
                        model.fit(X_train_scaled, y_train)
                    else:
                        model.fit(X_train, y_train)
                    self.models[name] = model
                except Exception:
                    pass
            
            # Don't print success message to avoid JSON parsing issues
            
        except Exception:
            # Don't print errors to avoid JSON parsing issues
            pass
    
    def predict_readmission(self, patient_data: Dict[str, Any]) -> Dict[str, Any]:
        """Predict readmission risk for a patient"""
        try:
            # Calculate clinical score
            clinical_score, risk_factors = self._calculate_clinical_score(patient_data)
            
            # Get ML prediction if available
            ml_probability = 0.25  # Default
            if ML_AVAILABLE and self.models:
                ml_probability = self._get_ml_prediction(patient_data)
            
            # Age-based adjustment
            age = patient_data.get('age', 50)
            age_group = self._get_age_group(age)
            age_multiplier = self.age_multipliers.get(age_group, 1.0)
            
            # Combine predictions
            base_probability = (0.6 * ml_probability) + (0.4 * clinical_score)
            final_probability = min(base_probability * age_multiplier, 1.0)
            
            # Critical condition overrides
            if self._has_critical_conditions(patient_data):
                final_probability = max(final_probability, 0.70)
                risk_factors.append('Critical medical conditions detected')
            
            risk_level = self._get_risk_level(final_probability)
            
            return {
                'patient_id': patient_data.get('patient_id', 'unknown'),
                'risk_level': risk_level,
                'risk_percentage': round(min(final_probability * 100, 100.0), 1),
                'confidence': round(self._calculate_confidence(ml_probability, clinical_score), 1),
                'risk_factors': risk_factors,
                'recommendation': self._get_recommendation(risk_level, final_probability),
                'age_group': age_group,
                'detailed_analysis': {
                    'primary_concerns': self._get_primary_concerns(risk_factors),
                    'preventive_measures': self._get_preventive_measures(risk_level),
                    'monitoring_requirements': self._get_monitoring_requirements(risk_level)
                }
            }
            
        except Exception as e:
            return {
                'patient_id': patient_data.get('patient_id', 'unknown'),
                'error': f"Prediction error: {str(e)}",
                'risk_level': 'error'
            }
    
    def _get_ml_prediction(self, patient_data: Dict[str, Any]) -> float:
        """Get ML model prediction"""
        if not self.models:
            return 0.25
        
        try:
            # Prepare features
            features = []
            feature_names = ['age', 'diabetes', 'hypertension', 'heart_disease', 'kidney_disease', 
                           'respiratory_disease', 'intensive_care_unit_admission', 'hemoglobin', 
                           'platelets', 'urea', 'length_of_stay', 'sars_cov2_exam_result', 
                           'previous_admissions', 'num_medications']
            
            for feature in feature_names:
                value = patient_data.get(feature, 0)
                try:
                    features.append(float(value))
                except:
                    features.append(0.0)
            
            features_array = np.array(features).reshape(1, -1)
            
            # Get predictions from all models
            predictions = []
            for name, model in self.models.items():
                try:
                    if name == 'logistic_regression' and 'standard' in self.scalers:
                        features_scaled = self.scalers['standard'].transform(features_array)
                        prob = model.predict_proba(features_scaled)[0, 1]
                    else:
                        prob = model.predict_proba(features_array)[0, 1]
                    predictions.append(prob)
                except:
                    predictions.append(0.25)
            
            return np.mean(predictions) if predictions else 0.25
            
        except Exception as e:
            print(f"ML prediction error: {e}")
            return 0.25
    
    def _calculate_clinical_score(self, patient_data: Dict[str, Any]) -> tuple:
        """Calculate clinical risk score"""
        risk_score = 0.0
        risk_factors = []
        
        # Age factors
        age = patient_data.get('age', 50)
        if age >= 80:
            risk_score += self.clinical_weights['age_elderly']
            risk_factors.append('Advanced age (80+ years) - very high risk')
        elif age >= 65:
            risk_score += self.clinical_weights['age_senior']
            risk_factors.append('Senior age (65-80 years) - increased risk')
        
        # Chronic conditions
        conditions = {
            'diabetes': 'Diabetes mellitus',
            'hypertension': 'Hypertension',
            'heart_disease': 'Heart disease',
            'kidney_disease': 'Kidney disease',
            'respiratory_disease': 'Respiratory disease'
        }
        
        active_conditions = []
        for condition, label in conditions.items():
            if patient_data.get(condition, 0) == 1:
                risk_score += self.clinical_weights.get(condition, 0.08)
                active_conditions.append(label)
        
        if len(active_conditions) >= 3:
            risk_score += self.clinical_weights['multiple_comorbidities']
            risk_factors.append(f'Multiple comorbidities: {", ".join(active_conditions)}')
        elif active_conditions:
            risk_factors.extend(active_conditions)
        
        # ICU admission
        if patient_data.get('intensive_care_unit_admission', 0) == 1:
            risk_score += self.clinical_weights['icu_admission']
            risk_factors.append('ICU admission - critical condition')
        
        # Lab values
        hb = patient_data.get('hemoglobin', 13.0)
        if hb < 12 and hb > 0:
            risk_score += self.clinical_weights['low_hemoglobin']
            risk_factors.append(f'Low hemoglobin (anemia): {hb}')
        
        platelets = patient_data.get('platelets', 250.0)
        if platelets < 150 and platelets > 0:
            risk_score += self.clinical_weights['low_platelets']
            risk_factors.append(f'Low platelets (thrombocytopenia): {platelets}')
        
        urea = patient_data.get('urea', 5.0)
        if urea > 7.5:
            risk_score += self.clinical_weights['high_urea']
            risk_factors.append(f'Elevated urea (kidney dysfunction): {urea}')
        
        # COVID-19
        if patient_data.get('sars_cov2_exam_result', 0) == 1:
            risk_score += self.clinical_weights['covid_positive']
            risk_factors.append('COVID-19 positive')
        
        # Length of stay
        los = patient_data.get('length_of_stay', 5)
        if los > 10:
            risk_score += self.clinical_weights['long_stay']
            risk_factors.append(f'Extended hospitalization ({los} days)')
        
        # Previous admissions
        prev_admissions = patient_data.get('previous_admissions', 0)
        if prev_admissions >= 2:
            risk_score += self.clinical_weights['frequent_admissions']
            risk_factors.append(f'Frequent admissions ({prev_admissions} in past year)')
        
        # High medication count
        medications = patient_data.get('num_medications', 5)
        if medications >= 10:
            risk_score += self.clinical_weights['high_medications']
            risk_factors.append(f'Polypharmacy ({medications} medications)')
        
        return min(risk_score, 1.0), risk_factors
    
    def _get_age_group(self, age: float) -> str:
        """Get age group from age"""
        if age < 35:
            return 'Young_Adult'
        elif age < 50:
            return 'Middle_Adult'
        elif age < 65:
            return 'Mature_Adult'
        elif age < 80:
            return 'Senior'
        else:
            return 'Elderly'
    
    def _has_critical_conditions(self, patient_data: Dict[str, Any]) -> bool:
        """Check for critical conditions"""
        critical_indicators = [
            patient_data.get('intensive_care_unit_admission', 0) == 1,
            patient_data.get('hemoglobin', 13) < 8,
            patient_data.get('platelets', 250) < 50,
            patient_data.get('urea', 5) > 20,
            patient_data.get('length_of_stay', 5) > 20
        ]
        return any(critical_indicators)
    
    def _get_risk_level(self, probability: float) -> str:
        """Get risk level from probability"""
        if probability >= 0.55:
            return 'high'
        elif probability >= 0.25:
            return 'medium'
        else:
            return 'low'
    
    def _get_recommendation(self, risk_level: str, probability: float) -> str:
        """Get clinical recommendations"""
        if risk_level == 'high':
            if probability >= 0.75:
                return "URGENT: Comprehensive discharge planning, intensive case management, transitional care team, 24h post-discharge contact, consider delaying discharge"
            else:
                return "HIGH RISK: Comprehensive discharge planning, case management, transitional care, 24-48h post-discharge contact"
        elif risk_level == 'medium':
            return "MODERATE RISK: Enhanced discharge planning, medication reconciliation, patient education, 3-7 day follow-up"
        else:
            return "LOW RISK: Standard discharge planning, routine follow-up in 1-2 weeks, patient education materials"
    
    def _calculate_confidence(self, ml_probability: float, clinical_score: float) -> float:
        """Calculate prediction confidence"""
        base_conf = 0.75
        ml_bonus = min(abs(ml_probability - 0.5) * 0.4, 0.15)  # Higher confidence when probability is further from 0.5
        clinical_bonus = min(clinical_score * 0.15, 0.15)
        return min(base_conf + ml_bonus + clinical_bonus, 1.0) * 100

    def _get_primary_concerns(self, risk_factors: list) -> list:
        """Get primary medical concerns from risk factors"""
        primary_concerns = []
        if any('age' in factor.lower() for factor in risk_factors):
            primary_concerns.append('Age-related increased vulnerability')
        if any('icu' in factor.lower() or 'critical' in factor.lower() for factor in risk_factors):
            primary_concerns.append('Critical care requirements')
        if any('comorbid' in factor.lower() or 'multiple' in factor.lower() for factor in risk_factors):
            primary_concerns.append('Complex medical conditions')
        if any('covid' in factor.lower() for factor in risk_factors):
            primary_concerns.append('COVID-19 complications')
        return primary_concerns or ['General medical monitoring required']

    def _get_preventive_measures(self, risk_level: str) -> list:
        """Get preventive measures based on risk level"""
        if risk_level == 'high':
            return [
                'Intensive discharge planning with care team',
                'Home health services arrangement',
                'Medication reconciliation and education',
                'Early follow-up appointment scheduling',
                'Emergency contact protocols'
            ]
        elif risk_level == 'medium':
            return [
                'Enhanced patient education materials',
                'Medication review and adherence plan',
                'Follow-up appointment within one week',
                'Clear discharge instructions',
                'Emergency warning signs education'
            ]
        else:
            return [
                'Standard discharge education',
                'Routine follow-up scheduling',
                'Basic medication instructions',
                'Emergency contact information'
            ]

    def _get_monitoring_requirements(self, risk_level: str) -> list:
        """Get monitoring requirements based on risk level"""
        if risk_level == 'high':
            return [
                'Daily check-ins for first 48-72 hours',
                'Weekly clinical assessments',
                'Medication compliance monitoring',
                'Vital signs tracking',
                'Symptoms progression monitoring'
            ]
        elif risk_level == 'medium':
            return [
                'Follow-up within 3-7 days',
                'Bi-weekly progress checks',
                'Medication effectiveness review',
                'Symptoms monitoring'
            ]
        else:
            return [
                'Routine follow-up in 1-2 weeks',
                'Standard care protocols',
                'As-needed consultations'
            ]

# Create global service instance
ml_service = SimplifiedMLService()
